File: src/main/java/backend/model/user/Participant.java
package backend.model.user;

import javax.persistence.DiscriminatorValue;
import javax.persistence.Entity;

@Entity
@DiscriminatorValue("PARTICIPANT")
public class Participant extends UserRole {

    private String emergencynumber;
    private String tshirtsize;
    private String hometown;
    private String phonenumber;

    public Participant() {
        super();
    }
    public Participant(UserCore core) {
        super(core);
    }

    public String getEmergencynumber() {
        return emergencynumber;
    }

    public void setEmergencynumber(String emergencynumber) {
        this.emergencynumber = emergencynumber;
    }

    public String getTshirtsize() {
        return tshirtsize;
    }

    public void setTshirtsize(String tshirtsize) {
        this.tshirtsize = tshirtsize;
    }

    public String getHometown() {
        return hometown;
    }

    public void setHometown(String hometown) {
        this.hometown = hometown;
    }

    public String getPhonenumber() {
        return phonenumber;
    }

    public void setPhonenumber(String phonenumber) {
        this.phonenumber = phonenumber;
    }

    @Override
    public String getAuthority() {
        return "PARTICIPANT";
    }
}

File: src/main/java/backend/model/user/Participant.kt
@file:JvmName("Participant")

package backend.model.user

import backend.exceptions.DomainException
import backend.model.event.Team
import backend.model.location.Location
import java.time.LocalDate
import java.util.*
import javax.persistence.*

@Entity
@DiscriminatorValue("PARTICIPANT")
class Participant : UserRole {

    var emergencynumber: String = ""
    var tshirtsize: String? = null
    var hometown: String? = null
    var phonenumber: String? = null
    var birthdate: LocalDate? = null

    @ManyToOne
    private var currentTeam: Team? = null

    @ManyToMany
    private var teams: MutableSet<Team>

    @OneToMany(mappedBy = "uploader")
    val locations: MutableList<Location> = ArrayList()

    /**
     * Private constructor for JPA
     */
    private constructor() : super() {
        teams = mutableSetOf()
    }

    constructor(account: UserAccount) : super(account) {
        teams = mutableSetOf()
    }

    fun getCurrentTeam(): Team? {
        return this.currentTeam
    }

    fun getAllTeams(): Set<Team> {
        return this.teams
    }

    fun clearAllTeams() {
        this.currentTeam = null
        this.teams.clear()
    }

    fun setCurrentTeam(team: Team?) {
        if (team != null) {
            this.teams.add(team)
        }
        this.currentTeam = team
    }

    override fun getAuthority(): String {
        return "PARTICIPANT"
    }

    @PreRemove
    fun preRemove() {
        this.locations.forEach { it.uploader = null }
        this.locations.clear()
    }

    fun removeTeam(team: Team) {
        if (this.getCurrentTeam() == team) {
            this.setCurrentTeam(null)
        }

        if (!this.teams.remove(team)) {
            throw DomainException("Can't remove team ${team.id} from participant $id. " +
                    "The participant has never been a part of this team")
        }
    }
}
